File: bindings/binding_typescript_wasm/__tests__/transform.js
const swc = require("../pkg");
const fs = require("node:fs");
const path = require("node:path");

function validatedImport(specifier, exports) {
    let out = `__nodeREPLDynamicImport(${JSON.stringify(specifier)})`;

    if (exports.length > 0) {
        out += `.then((m) => { `;
        for (const exportName of exports) {
            const message = `The requested module '${specifier}' does not provide an export named '${exportName}'`;
            out += `if (!(${JSON.stringify(exportName)} in m)) throw new SyntaxError(${JSON.stringify(message)}); `;
        }
        out += `return m; })`;
    }

    return out;
}

it("properly reports error", () => {
    expect(() => {
        swc.transformSync("Foo {}", {});
    }).toThrow();
});

describe("transform", () => {
    it("should strip types", async () => {
        const code = await swc.transform(
            `
        export const foo: number = 1;
        type Foo = number;
    `,
            {},
        );
        expect(code).toMatchSnapshot();
    });

    describe("in strip-only mode", () => {
        it("should remove declare enum", async () => {
            await expect(
                swc.transform(`declare enum Foo {}`, {}),
            ).resolves.toMatchSnapshot();
            await expect(
                swc.transform(
                    `declare enum Foo {
                    A
                }`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
            expect(
                swc.transform(
                    `declare enum Foo {
                    a = 2,
                    b,
                    }`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should strip type declarations", async () => {
            await expect(
                swc.transform(
                    `const foo = 1;
                    type Foo = number;
                    type Bar = string;
                    const bar: Bar = "bar";`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should strip type annotations", async () => {
            await expect(
                swc.transform(
                    `const foo = 1;
                    const bar: Bar = "bar";`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should strip type assertions", async () => {
            await expect(
                swc.transform(
                    `const foo = 1 as number;
                    const bar = "bar";`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should strip nonnull assertions", async () => {
            await expect(
                swc.transform(
                    `const foo = 1!;
                    const bar = "bar";`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should strip satisfies", async () => {
            await expect(
                swc.transform(
                    `const foo = 1 satisfies number;
                    const bar = "bar";`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should strip complex expressions", async () => {
            await expect(
                swc.transform(
                    `const foo = {
                        foo: 1 as number,
                        bar: "bar" as any as number,
                    } satisfies number;
                    const bar = "bar";`,
                    {},
                ),
            ).resolves.toMatchSnapshot();
        });

        it("should throw an error when it encounters an enum", async () => {
            await expect(
                swc.transform("enum Foo {}", {
                    mode: "strip-only",
                    filename: "test.ts",
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should throw an error when it encounters a namespace", async () => {
            await expect(
                swc.transform("namespace Foo { export const m = 1; }", {
                    mode: "strip-only",
                    filename: "test.ts",
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should throw an error when it encounters a module", async () => {
            await expect(
                swc.transform("module foo { }", {
                    mode: "strip-only",
                    deprecatedTsModuleAsError: true,
                    filename: "test.ts",
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should throw an error when it encounters a module", async () => {
            await expect(
                swc.transform("declare module foo { }", {
                    mode: "strip-only",
                    deprecatedTsModuleAsError: true,
                    filename: "test.ts",
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should not emit 'Caused by: failed to parse'", async () => {
            await expect(
                swc.transform("function foo() { await Promise.resolve(1); }", {
                    filename: "test.ts",
                    mode: "strip-only",
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should report correct error for syntax error", async () => {
            await expect(
                swc.transform("function foo() { invalid syntax }", {
                    mode: "strip-only",
                    filename: "test.ts"
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should report correct error for unsupported syntax", async () => {
            await expect(
                swc.transform(
                    `enum Foo {
                    a, b    
                    }`,
                    {
                        mode: "strip-only",
                        filename: "test.ts"
                    },
                ),
            ).rejects.toMatchSnapshot();
        });

        it("should not panic memory access out of bounds", async () => {
            expect(() => swc.transformSync(`
                var data = ""
if(!(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((data === "AED") || (data === "AFN")) || (data === "ALL")) || (data === "AMD")) || (data === "ANG")) || (data === "AOA")) || (data === "ARS")) || (data === "AUD")) || (data === "AWG")) || (data === "AZN")) || (data === "BAM")) || (data === "BBD")) || (data === "BDT")) || (data === "BGN")) || (data === "BHD")) || (data === "BIF")) || (data === "BMD")) || (data === "BND")) || (data === "BOB")) || (data === "BOV")) || (data === "BRL")) || (data === "BSD")) || (data === "BTN")) || (data === "BWP")) || (data === "BYN")) || (data === "BZD")) || (data === "CAD")) || (data === "CDF")) || (data === "CHE")) || (data === "CHF")) || (data === "CHW")) || (data === "CLF")) || (data === "CLP")) || (data === "CNY")) || (data === "COP")) || (data === "COU")) || (data === "CRC")) || (data === "CUC")) || (data === "CUP")) || (data === "CVE")) || (data === "CZK")) || (data === "DJF")) || (data === "DKK")) || (data === "DOP")) || (data === "DZD")) || (data === "EGP")) || (data === "ERN")) || (data === "ETB")) || (data === "EUR")) || (data === "FJD")) || (data === "FKP")) || (data === "GBP")) || (data === "GEL")) || (data === "GHS")) || (data === "GIP")) || (data === "GMD")) || (data === "GNF")) || (data === "GTQ")) || (data === "GYD")) || (data === "HKD")) || (data === "HNL")) || (data === "HRK")) || (data === "HTG")) || (data === "HUF")) || (data === "IDR")) || (data === "ILS")) || (data === "INR")) || (data === "IQD")) || (data === "IRR")) || (data === "ISK")) || (data === "JMD")) || (data === "JOD")) || (data === "JPY")) || (data === "KES")) || (data === "KGS")) || (data === "KHR")) || (data === "KMF")) || (data === "KPW")) || (data === "KRW")) || (data === "KWD")) || (data === "KYD")) || (data === "KZT")) || (data === "LAK")) || (data === "LBP")) || (data === "LKR")) || (data === "LRD")) || (data === "LSL")) || (data === "LYD")) || (data === "MAD")) || (data === "MDL")) || (data === "MGA")) || (data === "MKD")) || (data === "MMK")) || (data === "MNT")) || (data === "MOP")) || (data === "MRU")) || (data === "MUR")) || (data === "MVR")) || (data === "MWK")) || (data === "MXN")) || (data === "MXV")) || (data === "MYR")) || (data === "MZN")) || (data === "NAD")) || (data === "NGN")) || (data === "NIO")) || (data === "NOK")) || (data === "NPR")) || (data === "NZD")) || (data === "OMR")) || (data === "PAB")) || (data === "PEN")) || (data === "PGK")) || (data === "PHP")) || (data === "PKR")) || (data === "PLN")) || (data === "PYG")) || (data === "QAR")) || (data === "RON")) || (data === "RSD")) || (data === "RUB")) || (data === "RWF")) || (data === "SAR")) || (data === "SBD")) || (data === "SCR")) || (data === "SDG")) || (data === "SEK")) || (data === "SGD")) || (data === "SHP")) || (data === "SLL")) || (data === "SOS")) || (data === "SRD")) || (data === "SSP")) || (data === "STN")) || (data === "SVC")) || (data === "SYP")) || (data === "SZL")) || (data === "THB")) || (data === "TJS")) || (data === "TMT")) || (data === "TND")) || (data === "TOP")) || (data === "TRY")) || (data === "TTD")) || (data === "TWD")) || (data === "TZS")) || (data === "UAH")) || (data === "UGX")) || (data === "USD")) || (data === "USN")) || (data === "UYI")) || (data === "UYU")) || (data === "UYW")) || (data === "UZS")) || (data === "VES")) || (data === "VND")) || (data === "VUV")) || (data === "WST")) || (data === "XAF")) || (data === "XAG")) || (data === "XAU")) || (data === "XBA")) || (data === "XBB")) || (data === "XBC")) || (data === "XBD")) || (data === "XCD")) || (data === "XDR")) || (data === "XOF")) || (data === "XPD")) || (data === "XPF")) || (data === "XPT")) || (data === "XSU")) || (data === "XTS")) || (data === "XUA")) || (data === "XXX")) || (data === "YER")) || (data === "ZAR")) || (data === "ZMW")) || (data === "ZWL"))) {
  console.log("Hello");
}    
            `)).not.toThrow();
        })
    });

    describe("in transform mode", () => {
        it("should throw an error when it encounters a module", async () => {
            await expect(
                swc.transform("module foo { }", {
                    mode: "transform",
                    deprecatedTsModuleAsError: true,
                    filename: "test.ts"
                }),
            ).rejects.toMatchSnapshot();
        });

        it("should throw an error when it encounters a declared module", async () => {
            await expect(
                swc.transform("declare module foo { }", {
                    mode: "transform",
                    deprecatedTsModuleAsError: true,
                    filename: "test.ts",
                }),
            ).rejects.toMatchSnapshot();
        });

        it('shoud throw an object even with deprecatedTsModuleAsError = true', async () => {
            await expect(
                swc.transform("module F { export type x = number }", {
                    mode: "transform",
                    deprecatedTsModuleAsError: true,
                }),
            ).rejects.toMatchSnapshot();
        })
    });
});

describe("nodejs namespace", () => {
    it("exposes Node.js-specific helpers only through the namespace", () => {
        expect(swc.nodejs).toBeDefined();
        expect(typeof swc.nodejs.transformModuleSyntax).toBe("function");
        expect(typeof swc.__nodejsTransformModuleSyntax).toBe("undefined");
    });

    it("transforms module syntax without exposing an AST", () => {
        expect(
            swc.nodejs.transformModuleSyntax(`import fs from "node:fs";`),
        ).toEqual({
            code: `const { default: fs } = await ${validatedImport("node:fs", ["default"])};`,
            hadModuleSyntax: true,
        });

        expect(
            swc.nodejs.transformModuleSyntax(`const importName = "import";`),
        ).toEqual({
            code: `const importName = "import";`,
            hadModuleSyntax: false,
        });
    });

    it("supports Uint8Array input", () => {
        const input = new TextEncoder().encode(`import "node:fs";`);

        expect(swc.nodejs.transformModuleSyntax(input)).toEqual({
            code: `await __nodeREPLDynamicImport("node:fs");`,
            hadModuleSyntax: true,
        });
    });

    it("extracts the first expression for assertion locations", () => {
        expect(swc.nodejs.getFirstExpression("assert.ok(value)", 9)).toBe(
            "assert.ok(value)",
        );
        expect(swc.nodejs.getFirstExpression("assert(value)", 6)).toBe(
            "assert(value)",
        );
        expect(swc.nodejs.getFirstExpression("a(); assert.ok(value); b()", 13)).toBe(
            "assert.ok(value)",
        );
    });

    it("checks syntax validity and recoverability", () => {
        expect(swc.nodejs.isValidSyntax("await value")).toBe(true);
        expect(swc.nodejs.isValidSyntax("const x: number = 1")).toBe(true);
        expect(swc.nodejs.isValidSyntax("function foo(")).toBe(false);
        expect(swc.nodejs.isRecoverableError("function foo() {")).toBe(true);
        expect(swc.nodejs.isRecoverableError("const x: number = 1")).toBe(false);
        expect(swc.nodejs.isRecoverableError("2e")).toBe(false);
    });

    it("declares the public namespace in generated types", () => {
        const types = fs.readFileSync(path.join(__dirname, "../pkg/wasm.d.ts"), "utf8");

        expect(types).toContain("export declare namespace nodejs");
        expect(types).toContain("function transformModuleSyntax");
        expect(types).not.toContain("__nodejsTransformModuleSyntax");
    });
});

File: crates/swc_ts_fast_strip_binding/tests/transform.rs
use swc_ecma_parser::TsSyntax;
use swc_ts_fast_strip::{Mode, Options, TransformConfig};
use swc_ts_fast_strip_binding::transform;

/// Helper function to create default options with a specific mode
fn opts(mode: Mode) -> Options {
    Options {
        module: None,
        filename: Some("test.ts".into()),
        parser: TsSyntax {
            decorators: true,
            ..Default::default()
        },
        mode,
        transform: Some(TransformConfig {
            typescript: swc_ecma_transforms_typescript::Config {
                native_class_properties: true,
                ..Default::default()
            },
        }),
        deprecated_ts_module_as_error: Some(true),
        ..Default::default()
    }
}

/// Helper function to create options without deprecated_ts_module_as_error
fn opts_no_filename(mode: Mode) -> Options {
    Options {
        module: None,
        filename: None,
        parser: TsSyntax {
            decorators: true,
            ..Default::default()
        },
        mode,
        transform: Some(TransformConfig {
            typescript: swc_ecma_transforms_typescript::Config {
                native_class_properties: true,
                ..Default::default()
            },
        }),
        deprecated_ts_module_as_error: Some(true),
        ..Default::default()
    }
}

#[test]
fn properly_reports_error() {
    let result = transform("Foo {}".into(), opts_no_filename(Mode::StripOnly));
    assert!(result.is_err(), "should return Err for invalid syntax");
}

#[test]
fn should_strip_types() {
    let code = r#"
        export const foo: number = 1;
        type Foo = number;
    "#;

    let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
    assert!(result.is_ok(), "should succeed");

    let output = result.unwrap();
    assert!(output.code.contains("export const foo"));
    assert!(!output.code.contains(": number"));
    assert!(!output.code.contains("type Foo"));
}

mod strip_only_mode {
    use super::*;

    #[test]
    fn should_remove_declare_enum_empty() {
        let result = transform(
            "declare enum Foo {}".into(),
            opts_no_filename(Mode::StripOnly),
        );
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("declare enum"));
        assert!(!output.code.contains("enum"));
    }

    #[test]
    fn should_remove_declare_enum_with_member() {
        let code = "declare enum Foo {\n    A\n}";
        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("declare enum"));
        assert!(!output.code.contains("enum"));
    }

    #[test]
    fn should_remove_declare_enum_with_values() {
        let code = "declare enum Foo {\n    a = 2,\n    b,\n}";
        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("declare enum"));
        assert!(!output.code.contains("enum"));
    }

    #[test]
    fn should_strip_type_declarations() {
        let code = r#"const foo = 1;
                    type Foo = number;
                    type Bar = string;
                    const bar: Bar = "bar";"#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(output.code.contains("const foo = 1"));
        assert!(output.code.contains("const bar"));
        assert!(!output.code.contains("type Foo"));
        assert!(!output.code.contains("type Bar"));
        assert!(!output.code.contains(": Bar"));
    }

    #[test]
    fn should_strip_type_annotations() {
        let code = r#"const foo = 1;
                    const bar: Bar = "bar";"#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains(": Bar"));
    }

    #[test]
    fn should_strip_type_assertions() {
        let code = r#"const foo = 1 as number;
                    const bar = "bar";"#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("as number"));
        assert!(output.code.contains("const foo = 1"));
    }

    #[test]
    fn should_strip_nonnull_assertions() {
        let code = r#"const foo = 1!;
                    const bar = "bar";"#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("1!"));
        assert!(output.code.contains("const foo = 1"));
    }

    #[test]
    fn should_strip_satisfies() {
        let code = r#"const foo = 1 satisfies number;
                    const bar = "bar";"#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("satisfies"));
        assert!(output.code.contains("const foo = 1"));
    }

    #[test]
    fn should_strip_complex_expressions() {
        let code = r#"const foo = {
                        foo: 1 as number,
                        bar: "bar" as any as number,
                    } satisfies number;
                    const bar = "bar";"#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        assert!(result.is_ok(), "should succeed");

        let output = result.unwrap();
        assert!(!output.code.contains("as number"));
        assert!(!output.code.contains("as any"));
        assert!(!output.code.contains("satisfies"));
    }

    #[test]
    fn should_throw_error_on_enum() {
        let result = transform("enum Foo {}".into(), opts(Mode::StripOnly));
        assert!(result.is_err(), "should return Err for enum");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_throw_error_on_namespace() {
        let result = transform(
            "namespace Foo { export const m = 1; }".into(),
            opts(Mode::StripOnly),
        );
        assert!(result.is_err(), "should return Err for namespace");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_throw_error_on_module() {
        let mut options = opts(Mode::StripOnly);
        options.deprecated_ts_module_as_error = Some(true);

        let result = transform("module foo { }".into(), options);
        assert!(result.is_err(), "should return Err for module");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_throw_error_on_declare_module() {
        let mut options = opts(Mode::StripOnly);
        options.deprecated_ts_module_as_error = Some(true);

        let result = transform("declare module foo { }".into(), options);
        assert!(result.is_err(), "should return Err for declare module");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_not_emit_caused_by_failed_to_parse() {
        let result = transform(
            "function foo() { await Promise.resolve(1); }".into(),
            opts(Mode::StripOnly),
        );
        assert!(result.is_err(), "should return Err for await outside async");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");

        // Check that the error message doesn't contain "Caused by: failed to parse"
        let error_messages: Vec<String> = err.iter().map(|d| d.message.clone()).collect();
        for msg in error_messages {
            assert!(
                !msg.contains("Caused by: failed to parse"),
                "should not contain 'Caused by: failed to parse'"
            );
        }
    }

    #[test]
    fn should_report_correct_error_for_syntax_error() {
        let result = transform(
            "function foo() { invalid syntax }".into(),
            opts(Mode::StripOnly),
        );
        assert!(result.is_err(), "should return Err for syntax error");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
        assert!(err[0].message.len() > 0, "should have error message");
    }

    #[test]
    fn should_report_correct_error_for_unsupported_syntax() {
        let code = "enum Foo {\n    a, b    \n}";
        let result = transform(code.into(), opts(Mode::StripOnly));
        assert!(result.is_err(), "should return Err for unsupported enum");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_not_panic_memory_access_out_of_bounds() {
        let code = r#"
                var data = ""
if(!(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((data === "AED") || (data === "AFN")) || (data === "ALL")) || (data === "AMD")) || (data === "ANG")) || (data === "AOA")) || (data === "ARS")) || (data === "AUD")) || (data === "AWG")) || (data === "AZN")) || (data === "BAM")) || (data === "BBD")) || (data === "BDT")) || (data === "BGN")) || (data === "BHD")) || (data === "BIF")) || (data === "BMD")) || (data === "BND")) || (data === "BOB")) || (data === "BOV")) || (data === "BRL")) || (data === "BSD")) || (data === "BTN")) || (data === "BWP")) || (data === "BYN")) || (data === "BZD")) || (data === "CAD")) || (data === "CDF")) || (data === "CHE")) || (data === "CHF")) || (data === "CHW")) || (data === "CLF")) || (data === "CLP")) || (data === "CNY")) || (data === "COP")) || (data === "COU")) || (data === "CRC")) || (data === "CUC")) || (data === "CUP")) || (data === "CVE")) || (data === "CZK")) || (data === "DJF")) || (data === "DKK")) || (data === "DOP")) || (data === "DZD")) || (data === "EGP")) || (data === "ERN")) || (data === "ETB")) || (data === "EUR")) || (data === "FJD")) || (data === "FKP")) || (data === "GBP")) || (data === "GEL")) || (data === "GHS")) || (data === "GIP")) || (data === "GMD")) || (data === "GNF")) || (data === "GTQ")) || (data === "GYD")) || (data === "HKD")) || (data === "HNL")) || (data === "HRK")) || (data === "HTG")) || (data === "HUF")) || (data === "IDR")) || (data === "ILS")) || (data === "INR")) || (data === "IQD")) || (data === "IRR")) || (data === "ISK")) || (data === "JMD")) || (data === "JOD")) || (data === "JPY")) || (data === "KES")) || (data === "KGS")) || (data === "KHR")) || (data === "KMF")) || (data === "KPW")) || (data === "KRW")) || (data === "KWD")) || (data === "KYD")) || (data === "KZT")) || (data === "LAK")) || (data === "LBP")) || (data === "LKR")) || (data === "LRD")) || (data === "LSL")) || (data === "LYD")) || (data === "MAD")) || (data === "MDL")) || (data === "MGA")) || (data === "MKD")) || (data === "MMK")) || (data === "MNT")) || (data === "MOP")) || (data === "MRU")) || (data === "MUR")) || (data === "MVR")) || (data === "MWK")) || (data === "MXN")) || (data === "MXV")) || (data === "MYR")) || (data === "MZN")) || (data === "NAD")) || (data === "NGN")) || (data === "NIO")) || (data === "NOK")) || (data === "NPR")) || (data === "NZD")) || (data === "OMR")) || (data === "PAB")) || (data === "PEN")) || (data === "PGK")) || (data === "PHP")) || (data === "PKR")) || (data === "PLN")) || (data === "PYG")) || (data === "QAR")) || (data === "RON")) || (data === "RSD")) || (data === "RUB")) || (data === "RWF")) || (data === "SAR")) || (data === "SBD")) || (data === "SCR")) || (data === "SDG")) || (data === "SEK")) || (data === "SGD")) || (data === "SHP")) || (data === "SLL")) || (data === "SOS")) || (data === "SRD")) || (data === "SSP")) || (data === "STN")) || (data === "SVC")) || (data === "SYP")) || (data === "SZL")) || (data === "THB")) || (data === "TJS")) || (data === "TMT")) || (data === "TND")) || (data === "TOP")) || (data === "TRY")) || (data === "TTD")) || (data === "TWD")) || (data === "TZS")) || (data === "UAH")) || (data === "UGX")) || (data === "USD")) || (data === "USN")) || (data === "UYI")) || (data === "UYU")) || (data === "UYW")) || (data === "UZS")) || (data === "VES")) || (data === "VND")) || (data === "VUV")) || (data === "WST")) || (data === "XAF")) || (data === "XAG")) || (data === "XAU")) || (data === "XBA")) || (data === "XBB")) || (data === "XBC")) || (data === "XBD")) || (data === "XCD")) || (data === "XDR")) || (data === "XOF")) || (data === "XPD")) || (data === "XPF")) || (data === "XPT")) || (data === "XSU")) || (data === "XTS")) || (data === "XUA")) || (data === "XXX")) || (data === "YER")) || (data === "ZAR")) || (data === "ZMW")) || (data === "ZWL"))) {
  console.log("Hello");
}
            "#;

        let result = transform(code.into(), opts_no_filename(Mode::StripOnly));
        // Should not panic - either succeed or return an error
        assert!(
            result.is_ok() || result.is_err(),
            "should not panic on deeply nested expression"
        );
    }
}

mod transform_mode {
    use super::*;

    #[test]
    fn should_throw_error_on_module() {
        let mut options = opts(Mode::Transform);
        options.deprecated_ts_module_as_error = Some(true);

        let result = transform("module foo { }".into(), options);
        assert!(result.is_err(), "should return Err for module");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_throw_error_on_declared_module() {
        let mut options = opts(Mode::Transform);
        options.deprecated_ts_module_as_error = Some(true);

        let result = transform("declare module foo { }".into(), options);
        assert!(result.is_err(), "should return Err for declare module");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }

    #[test]
    fn should_throw_object_even_with_deprecated_ts_module_as_error() {
        let mut options = opts_no_filename(Mode::Transform);
        options.deprecated_ts_module_as_error = Some(true);

        let result = transform("module F { export type x = number }".into(), options);
        assert!(result.is_err(), "should return Err for module");

        let err = result.unwrap_err();
        assert!(!err.is_empty(), "should have error diagnostics");
    }
}
